resources/assets/lib/beatmap-discussions/new-reply.tsx

```diff
@@ -0,0 +1,272 @@
+import core from 'osu-core-singleton';
+import * as React from 'react';
+import TextareaAutosize from 'react-autosize-textarea';
+import { onError } from 'utils/ajax';
+import { validMessageLength } from 'utils/beatmapset-discussion-helper';
+import { InputEventType, makeTextAreaHandler, TextAreaCallback } from 'utils/input-handler';
+import { hideLoadingOverlay, showLoadingOverlay } from 'utils/loading-overlay';
+import MessageLengthCounter from './message-length-counter';
+
+const bn = 'beatmap-discussion-post';
+
+interface Props {
+  beatmapset: BeatmapsetJson;
+  currentBeatmap: BeatmapJson;
+  currentUser: CurrentUserJson;
+  discussion: BeatmapsetDiscussionJson & Required<Pick<BeatmapsetDiscussionJson, 'current_user_attributes'>>;
+}
+
+const actionIcons = {
+  reply: 'fas fa-reply',
+  reply_reopen: 'fas fa-exclamation-circle',
+  reply_resolve: 'fas fa-check',
+};
+
+const actionResolveLookup = {
+  reply_reopen: false,
+  reply_resolve: true,
+} as Record<string, boolean | undefined>;
+
+@observer
+export class NewReply extends React.Component<Props> {
+  @observable private readonly box = React.createRef<HTMLTextAreaElement>();
+  @observable private editing = osu.present(this.storedMessage);
+  private readonly handleKeyDown;
+  @observable private message = this.storedMessage;
+  @observable private posting: string | null = null;
+  private postXhr: JQuery.jqXHR<BeatmapsetDiscussionPostStoreResponseJson> | null = null;
+
+  private get canReopen() {
+    return this.props.discussion.can_be_resolved && this.props.discussion.current_user_attributes.can_reopen;
+  }
+
+  private get canResolve() {
+    return this.props.discussion.can_be_resolved && this.props.discussion.current_user_attributes.can_resolve;
+  }
+
+  private get isTimeline() {
+    return this.props.discussion.timestamp != null;
+  }
+
+  private get storageKey() {
+    return `beatmapset-discussion:reply:${this.props.discussion.id}:message`;
+  }
+
+  private get storedMessage() {
+    return localStorage.getItem(this.storageKey) ?? '';
+  }
+
+  private get validPost() {
+    return validMessageLength(this.message, this.isTimeline);
+  }
+
+  constructor(props: Props) {
+    super(props);
+    this.handleKeyDown = makeTextAreaHandler(this.handleKeyDownCallback);
+
+    makeObservable(this);
+  }
+
+  componentDidUpdate(prevProps: Readonly<Props>) {
+    if (prevProps.discussion.id !== this.props.discussion.id) {
+      this.message = this.storedMessage;
+      return;
+    }
+
+    if (this.editing) {
+      this.box.current?.focus();
+    }
+
+    this.storeMessage();
+  }
+
+  componentWillUnmount() {
+    this.postXhr?.abort();
+  }
+
+  render() {
+    return this.editing ? this.renderBox() : this.renderPlaceholder();
+  }
+
+  @action
+  private readonly editStart = () => {
+    if (core.userLogin.showIfGuest(this.editStart)) return;
+    this.editing = true;
+  };
+
+  private handleKeyDownCallback: TextAreaCallback = (type, event) => {
+    runInAction(() => {
```

Comment:
declaring `@action` isn't enough

Reply:
I think you just need to set `handleKeyDown` after `makeObservable` since that's when decorations are applied

Reply:
makes sense 🤔

Reply:
also missing readonly 👀